### DISCUSSION TITLE
Parse json object contains another multiple objects

### DISCUSSION THREAD

**Discussion by aser86** [ClickHouse/ClickHouse#60351] [Q&A] [answered]
Hello,
I have a json format which contains an object with other json objcets, see below Testfile.

JSON-File Example
{ "system": 123, "time" : "2023-10-01", "files": { "fileA": { "name": "myfile", "size": 118, "value": "text" }, "fileB": { "name": "myfile2", "size": 244, "value": "my text" }, "fileC": { "name": "myfile3", "size": 71, "value": "" } } }

JSON-File: [Testfile](https://github.com/aser86/test/blob/main/0_test_memory.json)

If i create an json file with only the object "files" this sql works fine:

`SELECT * FROM file('/bitnami/clickhouse/data/user_files/0_test_memory.json', JSONObjectEachRow) settings format_json_object_each_row_column_for_object_name='files';`

Then I tried to read the complete json format but i dont know how i can use it on a column, see the under sql.
Is there a possibility to do that? Maybe other ways are better?

WITH JSONExtract(json, '
Tuple(
	system Int64
	, time String
	, files String
)') AS parsed_json
, elements as (
	SELECT
    tupleElement(parsed_json, 'system') AS system
    , tupleElement(parsed_json, 'time') AS time
    , tupleElement(parsed_json, 'files') AS files
	from file('/bitnami/clickhouse/data/user_files/0_test_memory.json', 'JSONAsString', 'json String')
)
/*
, files as (
	SELECT * FROM (elements, JSONObjectEachRow, 'file_name String, name String, size String, value String')
	settings format_json_object_each_row_column_for_object_name='files'
)
*/
SELECT *
from elements;

My Goal is a table in this format:
system time files name size value
123 2023-10-01 fileA myfile 118 text
123 2023-10-01 fileB myfile2 244 my text
123 2023-10-01 fileC myfile3 71
...

ClickHouse version:
24.1.5.6

**Comment by mlex** [comment, marked as answer]
You can read the whole file with an explicit schema `'system Int64, time String, files String'` and then use the `JSONExtractKeysAndValuesRaw` function together with `ARRAY JOIN` to put every entry of the `files` key into a separate result row. The result can look like this:

```
SELECT
    system,
    time,
    files_raw.1 AS filekey,
    JSONExtractString(files_raw.2, 'name') AS filename,
    JSONExtractString(files_raw.2, 'size') AS filesize,
    JSONExtractString(files_raw.2, 'value') AS filevalue
FROM file('/tmp/0_test_memory.json', JSONEachRow, 'system Int64, time String, files String')
ARRAY JOIN JSONExtractKeysAndValuesRaw(files) AS files_raw
LIMIT 10

Query id: f0ca70fd-78ba-4e71-9a43-9e7b053d894d

┌─system─┬─time───────┬─filekey─┬─filename─┬─filesize─┬─filevalue─┐
│    123 │ 2023-10-01 │ fileA1  │ myfile1  │ 100      │ text1     │
│    123 │ 2023-10-01 │ fileA2  │ myfile2  │ 200      │ text2     │
│    123 │ 2023-10-01 │ fileA3  │ myfile3  │ 300      │ text3     │
│    123 │ 2023-10-01 │ fileA4  │ myfile4  │ 400      │ text4     │
│    123 │ 2023-10-01 │ fileA5  │ myfile5  │ 500      │ text5     │
│    123 │ 2023-10-01 │ fileA6  │ myfile6  │ 600      │ text6     │
│    123 │ 2023-10-01 │ fileA7  │ myfile7  │ 700      │ text7     │
│    123 │ 2023-10-01 │ fileA8  │ myfile8  │ 800      │ text8     │
│    123 │ 2023-10-01 │ fileA9  │ myfile9  │ 900      │ text9     │
│    123 │ 2023-10-01 │ fileA10 │ myfile10 │ 1000     │ text10    │
└────────┴────────────┴─────────┴──────────┴──────────┴───────────┘

10 rows in set. Elapsed: 0.002 sec.
```
